### ISSUE TITLE
Evaluate - Fine tuning the entire LLM network vs default classifier head vs bigger head

### ISSUE THREAD

**Issue by bdzyubak** [bdzyubak/torch-control#31]
The typical approach is to freeze all layers except for a small default head e.g. 768 channels for distilBERT. For this issue, test how does training the full network (assuming it fits in memory) compare to training the default head to a larger custom head. 

Metrics: 
validation accuracy
training accuracy 
training time

**Comment by bdzyubak** [comment]
Training of the full network takes 46 minutes and yields 0.88 validation accuracy on binarized positive/negative sentiments
projects\NaturalLanguageProcessing\MovieReviewAnalysis\fine_tune_on_kaggle_movie_sentiment.py
Commit: f0367f3c76cf38c790af43c25abc059d00557ba9

![image](https://github.com/bdzyubak/torch-control/assets/37943739/736842c1-9b94-4759-bec6-9b37d856ae1e)

**Comment by bdzyubak** [comment]
Training a larger classifier head with 768x768x768x1 parameters takes the same amount of time (other processes were running in parallel, doubtful it is actually slower than training the full network) and yields inferior accuracy to the full network. 

train_acc 83.7%
val_acc 83.4%
time: 50.7 min

![image](https://github.com/bdzyubak/torch-control/assets/37943739/bf236eca-50db-40c3-9600-4e5cf94331b4)

**Comment by bdzyubak** [comment]
Adding 10 additional layers does not improve accuracy. Variability between reruns appears to be about 2%
![image](https://github.com/bdzyubak/torch-control/assets/37943739/09c18d0d-e046-4c5d-bed7-5a479f5b7b80)

**Comment by bdzyubak** [comment]
Overall, training with the original fine tune head provides the same accuracy as increasing the size of the head by a factor of 10x to 3M. Training the entire network with 60 M parameters provides the highest train accuracy, and highest val accuracy due to checkpointing successfully dealing with overfitting. 

**Conclusion**: train the full network with checkpointing on val accuracy. 

Note: surprising that fine tuning with 300k and 6 M parameters in multiple linear layers with activation and 0 dropout results in the same 0.83-0.85 accuracy, while training the full network does improve train accuracy to 0.98. This appears to indicate that some weights in the initial layers are wrong and need to be updated for optimal performance, while having pretrained looses input information to where it is not recoverable by deeper layers.
